core/src/avm2/globals/flash/geom/Matrix.as

```diff
@@ -16,6 +19,103 @@ package flash.geom {
 			this.ty = ty;
 		}
 
+		public function clone():Matrix {
+			return new Matrix(this.a,this.b,this.c,this.d,this.tx,this.ty);
+		}
+
+		public function concat(m:Matrix):void {
+			var a = m.a * this.a + m.c * this.b;
+			var b = m.b * this.a + m.d * this.b;
+			var c = m.a * this.c + m.c * this.d;
+			var d = m.b * this.c + m.d * this.d;
+			var tx = m.a * this.tx + m.c * this.ty + m.tx;
+			var ty = m.b * this.tx + m.d * this.ty + m.ty;
+
+			this.a = a;
+			this.b = b;
+			this.c = c;
+			this.d = d;
+			this.tx = tx;
+			this.ty = ty;
+		}
+
+		public function copyColumnFrom(column:uint, vector3D:Vector3D):void {
+			// FP BUG: For some reason these methods are identical
+			this.copyRowFrom(column, vector3D);
+		}
+
+		public function copyColumnTo(column:uint, vector3D:Vector3D):void {
+			if(column == 0) {
+				vector3D.x = this.a;
+				vector3D.y = this.b;
+				vector3D.z = 0;
+			}
+			else if (column == 1) {
+				vector3D.x = this.c;
+				vector3D.y = this.d;
+				vector3D.z = 0;
+			}
+			else if (column == 2) {
+				vector3D.x = this.tx;
+				vector3D.y = this.ty;
+				vector3D.z = 1;
+			} // otherwise vector is unchanged
+		}
+
+		public function copyFrom(sourceMatrix: Matrix): void {
+			this.a = sourceMatrix.a;
+			this.b = sourceMatrix.b;
+			this.c = sourceMatrix.c;
+			this.d = sourceMatrix.d;
+			this.tx = sourceMatrix.tx;
+			this.ty = sourceMatrix.ty;
+		}
+
+		public function copyRowFrom(row: uint, vector3D: Vector3D): void {
+			if (row == 0) {
+				this.a = vector3D.x;
+				this.c = vector3D.y;
+				this.tx = vector3D.z;
+			} else if(row == 1) {
+				this.b = vector3D.x;
+				this.d = vector3D.y;
+				this.ty = vector3D.z;
+			} // otherwise matrix is unchanged
+		}
+
+		public function copyRowTo(row:uint, vector3D:Vector3D):void {
+			if(row == 0) {
+				vector3D.x = this.a;
+				vector3D.y = this.c;
+				vector3D.z = this.tx;
+			}
+			else if (row == 1) {
+				vector3D.x = this.b;
+				vector3D.y = this.d;
+				vector3D.z = this.ty;
+			}
+			else if (row == 2) {
+				vector3D.x = 0;
+				vector3D.y = 0;
+				vector3D.z = 1;
+			} // otherwise vector is unchanged
+		}
+
+		public function createBox(scaleX:Number, scaleY:Number, rotation:Number = 0, tx:Number = 0, ty:Number = 0):void {
+			this.identity();
+			this.rotate(rotation);
+			this.scale(scaleX, scaleY);
+			this.translate(tx, ty);
+		}
+
+		public function createGradientBox(width: Number, height: Number, rotation: Number = 0, tx: Number = 0, ty: Number = 0): void {
+			this.createBox(width / 1638.4, height / 1638.4, rotation, tx + width / 2, ty + height / 2);
```

Comment:
This magic number is a bit odd. Is this `2^(16-1) / Twips::TWIPS_PER_PIXEL`?

Reply:
Okay, in Rust, perhaps `32768 / Twips::TWIPS_PER_PIXEL` would be better.
Was this number simply observed, or is it documented somewhere?

Reply:
Never mind, this is AS3.